app/src/org/commcare/utils/FileUtil.java

```diff
@@ -553,22 +526,55 @@ public static String getExtension(String filePath) {
     /**
      * Retrieve a file's name using contentUri.
      * https://developer.android.com/training/secure-file-sharing/retrieve-info.html#RetrieveFileInfo
+     *
+     * @throws FileExtensionNotFoundException
      */
     public static String getFileName(Context context, Uri uri) {
+        String fileName;
         try (Cursor cursor = context.getContentResolver().query(uri, null, null, null, null)) {
             if (cursor == null || cursor.getCount() <= 0) {
                 try {
-                    return getFileName(UriToFilePath.getPathFromUri(context, uri));
+                    fileName = getFileName(UriToFilePath.getPathFromUri(context, uri));
                 } catch (UriToFilePath.NoDataColumnForUriException e) {
-                    return "";
+                    fileName = uri.getLastPathSegment();
+                }
+            } else {
+                int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
+                cursor.moveToFirst();
+                fileName = cursor.getString(nameIndex);
+                cursor.close();
+            }
+        }
+        if (TextUtils.isEmpty(getExtension(fileName))) {
+            String ext = getFileExtension(context, uri, fileName);
+            fileName = fileName + "." + ext;

```

Comment:
`.bin` or `.dat` would be more conventional choices for unknown binary extensions

Reply:
I'm sorry I didn't get this one. Do you mean if we aren't able to get the extension from the provider, we should use a `.bin` or `.dat` extension rather than throwing an exception?